flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/stream/DeduplicationTestPrograms.java

```diff
@@ -0,0 +1,151 @@
+public class DeduplicationTestPrograms {
+
+    static final Row[] DATA1 = {
+        Row.of(1L, "terry", "pen", 1000L),
+        Row.of(2L, "alice", "pen", 2000L),
+        Row.of(3L, "bob", "pen", 3000L),
+        Row.of(4L, "bob", "apple", 4000L),
+        Row.of(5L, "fish", "apple", 5000L)
+    };
+
+    static final Row[] DATA2 = {
+        Row.of(6L, "jerry", "pen", 6000L),
+        Row.of(7L, "larry", "apple", 7000L),
+        Row.of(8L, "bill", "banana", 8000L),
+        Row.of(9L, "carol", "apple", 9000L)
+    };
+
+    static final TableTestProgram DEDUPLICATE =
+            TableTestProgram.of("deduplicate-asc", "validates deduplication in ascending")
+                    .setupTableSource(
+                            SourceTestStep.newBuilder("MyTable")
+                                    .addSchema(
+                                            "order_id bigint",
+                                            "`user` varchar",
+                                            "product varchar",
+                                            "order_time bigint ",
+                                            "event_time as TO_TIMESTAMP(FROM_UNIXTIME(order_time)) ",
+                                            "watermark for event_time as event_time - INTERVAL '5' second ")
+                                    .producedBeforeRestore(DATA1)
+                                    .producedAfterRestore(DATA2)
+                                    .build())
+                    .setupTableSink(
+                            SinkTestStep.newBuilder("MySink")
+                                    .addSchema(
+                                            "order_id bigint",
+                                            "`user` varchar",
+                                            "product varchar",
+                                            "order_time bigint",
+                                            "primary key(product) not enforced")
+                                    .consumedBeforeRestore(
+                                            Row.of(1, "terry", "pen", 1000),
+                                            Row.of(4, "bob", "apple", 4000))
+                                    .consumedAfterRestore(Row.of(8L, "bill", "banana", 8000L))
+                                    .build())
+                    .runSql(
+                            "insert into MySink "
+                                    + "select order_id, user, product, order_time \n"
+                                    + "FROM ("
+                                    + "  SELECT *,"
+                                    + "    ROW_NUMBER() OVER (PARTITION BY product ORDER BY event_time ASC) AS row_num\n"
+                                    + "  FROM MyTable)"
+                                    + "WHERE row_num = 1")
+                    .build();
+
+    static final TableTestProgram DEDUPLICATE_DESC =
```

Comment:
I'm not completely sold on this test.  I added it to see an example with `Deduplicate(keep=[LastRow]...`.

If we want to remove it, I'm fine with that.  If the example needs an addition to make it make more sense, I'm open to suggestions.

Reply:
I like the idea of adding the test. The behaviour is so different that I think it's worth having a test for it.